### ISSUE TITLE
Critical error in autocorrect for `syntactic_sugar`

### ISSUE THREAD

**Issue by tsheaff** [realm/SwiftLint#3156]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

`swiftlint autocorrect` fails on the following code, in a way that fails to compile

```swift
func fooBarBaz() -> Dictionary<String, Dictionary<Int, Float>> {
    ...
}
```

It becomes:

```swift
func fooBarBaz() -> [String: Dictionary[Int: Float]> {
    ...
}
```

This does not compile.

It should become:

```swift
func fooBarBaz() -> [String: [Int: Float]] {
    ...
}
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
`0.39.1`

* Installation method used (Homebrew, CocoaPods, building from source, etc)?
CocoaPods

* Paste your configuration file:

```yml
disabled_rules: # rule identifiers to exclude from running
  - shorthand_operator
  - trailing_whitespace
  - line_length
  - cyclomatic_complexity
  - identifier_name
  - function_body_length
  - notification_center_detachment
  - multiple_closures_with_trailing_closure
  - file_length
  - type_body_length
  - switch_case_alignment
  - large_tuple
  - function_parameter_count
  - discouraged_direct_init
  - discarded_notification_center_observer
  - todo
  - no_fallthrough_only
  - nesting
  - operator_whitespace
  - type_name
  
opt_in_rules: # some rules are only opt-in
  - empty_count
  - empty_string
  - sorted_first_last
  - contains_over_first_not_nil
  - first_where
  - last_where
  - yoda_condition
  - prefer_self_type_over_type_of_self

included: # paths to include during linting. `--path` is ignored if present.
  - app/source
  - tests
  - watch/extension/source
  - widget

excluded: # paths to ignore during linting. Takes precedence over `included`.
  - Pods
  - fabric
  - fastlane
  - scripts
  - siri 

reporter: "xcode"
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcodebuild -version`)?

```
Xcode 11.0
Build version 11A419c
```

**Comment by wjling** [comment]
Same problem. It's critical X(

**Comment by mluisbrown** [comment]
The same problem occurs for optional closures:

```swift
let textDidChange: Optional<(String) -> Void>
```

becomes:
```swift
let textDidChange: (String) -? Void>
```

The problem seems to be that the regex is looking for the first closing `>`, not considering that this character may not be closing the generic type declaration.

**Comment by AndrewGable** [comment]
I believe we saw the same issue with this:

```swift
let completionHandler = { (placemarks: Optional<Array<CLPlacemark>>, error: Optional<Error>) in
```

being autocorrected to this:

```swift
let completionHandler = { (placemarks: ArrayCLPlacemark?>, error: Error?) in
```

**Comment by mluisbrown** [comment]
IMO this is a bug that should be fixed so I'm commenting to keep it open.

**Comment by irvinlesh-op** [comment]
Can confirm this is still an issue a year later. `Array<Array<N>>` gets turned into `[Array[N]>`...
